### DISCUSSION TITLE
MariaDB: Using a UUID as primary key

### DISCUSSION THREAD

**Discussion by noseshimself** [Budibase/budibase#7317] [Help] [answered]
I'm trying to do something rather exotic requiring the same mandatory fields for every table: created_on (timestamp, not null, default value current_timestamp()), updated_on (the same with a trigger on update) and GlobalID (UUID, default value uuid()). It seems that budibase doesn't even know the type timestamp...

Objects are defined like

```
CREATE TABLE `clients` (
  `GlobalID` varchar(36) NOT NULL DEFAULT uuid(),
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `created_at` timestamp NULL DEFAULT current_timestamp(),
  `updated_at` timestamp NULL DEFAULT current_timestamp() ON UPDATE current_timestamp(),
  PRIMARY KEY (`GlobalID`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
```

Is there any other way than defining the tables outside budibase and not touching these fields with an application at all or should I use another database engine where budibase knows the relevant field types? Can Budibase deal with triggers and functions at all?

**Comment by melohagan** [comment]
Hi @noseshimself 

I'm not sure what the issue is. I've tried the timestamp datatype on MariaDB and it appears to be working fine for myself:

<img width="247" alt="Screenshot 2022-08-18 at 15 22 14" src="https://user-images.githubusercontent.com/101575380/185418728-8ae1543f-9e6a-40e7-8702-c0667f95d5c0.png">

![Screenshot 2022-08-18 at 15 22 33](https://user-images.githubusercontent.com/101575380/185418795-8236c9c5-a1b6-436a-a20f-9b8624a74f6e.png)

**Reply by noseshimself** [reply]
> Hi @noseshimself
> 
> I'm not sure what the issue is. I've tried the timestamp datatype on MariaDB and it appears to be working fine for myself:

https://dev.mysql.com/doc/refman/8.0/en/datetime.html#:~:text=MySQL%20retrieves%20and%20displays%20DATETIME,%3A14%3A07'%20UTC.

A timestamp is not really the same as datetime (e. g. it is always kept in UTC, no matter what) and is totally painless for the purpose I intend it for (see my create table example above). As soon as something is creating or updating the row I'm getting things "notarized" by the back end even if some clever user is trying to fiddle with the data with his own client.

I would really like to be able to define database-inherent mechanisms from within Budibase (i. e. triggering of at least MySQL/mariadb built-in functions) although I can live with having to define my databases externally if budibase will at least show me that there are triggers attached to fields and recognizes structural changes applied externally.

But the main problem is still there: Having a simple way of using UUIDs as record ID instead of int (using the built-in types of databases if there is one -- Postgres would be one) and offering automatic packing (varchar (36) to 16 bytes binary representation) and unpacking would be a great thing to have.

I'm thinking of projects that need the inherent exchanging or even sharing of data between instances.

(I guess describing the intended application of these things would explain it a lot better but I'm not permitted to do that in public.)

**Comment by noseshimself** [comment, marked as answer]
I thought hard about the UUID-related part of my question. I probably thought too much about my existing solution and forgot that I now have an application server at my disposal which means I do not have the database server do things it is not too good at so please see feature request https://github.com/Budibase/budibase/issues/7391
